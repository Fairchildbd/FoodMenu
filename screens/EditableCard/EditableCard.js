import React, { useState } from 'react';
import { View, TouchableOpacity, Text } from 'react-native';
import { Card, Icon, Input } from 'react-native-elements';
import { menuStyles } from '../../styles/styles';
import testID from 'react-native-testid';

export function useEditing() {
  const [isEditing, setIsEditing] = useState(false);
  return { isEditing, setIsEditing };
}

export function EditableCard (props) {
  const [img, setImage] = useState('');
  const [name, setName] = useState('');
  const [description, setDescription] = useState('');
  const [price, setPrice] = useState('');
  const {isEditing, setIsEditing} = useEditing(false);
  const payload = {
    img,
    name,
    description,
    price,
  };
  const initialItem = {
    img: props?.item?.img,
    name: props?.item?.name,
    description: props?.item?.description,
    price: props?.item?.price
  };
  if(isEditing) {
    return (
      <View style={ menuStyles.container }>
        <Card containerStyle={ menuStyles.cardContainer }>
        <View style={ menuStyles.editButtonSet }>
          <TouchableOpacity { ...testID('Submission') } onPress={ () => { props.handleEditedItem({ editedItem: payload, initialItem, edited: true, }); setIsEditing(false); } }>
            <Icon name='done' color='white'/>
          </TouchableOpacity>
          <TouchableOpacity { ...testID('Cancel') } onPress={ () => { props.handleEditedItem(); setIsEditing(false) }}>
            <Icon name='close' color='white'/>
          </TouchableOpacity>
        </View>
        <Input
          { ...testID('Name') }
          placeholder={props?.item?.name}
          onChangeText={ value => setName(value) }
        />
        <Card.Divider width={2} color='white' />
        <Input
          { ...testID('Img') }
          placeholder={ props?.item?.img }
          onChangeText={ value => setImage(value) }
        />
        <Input
          placeholder={ props?.item?.description }
          onChangeText={ value => setDescription(value) }
        />
        <Input
          placeholder={ props?.item?.price }
          onChangeText={ value => setPrice(value) }
        />
      </Card>
    </View>
    );
  } else {
    return (
      <View style={ menuStyles.container }>
      <Card containerStyle={ menuStyles.cardContainer }>
        <View style={ { flexDirection: 'row', justifyContent: 'space-between' } }>
          <TouchableOpacity { ...testID('Edit') } onPress={ () => setIsEditing(true) }>
            <Icon name='edit' color='white'/>
          </TouchableOpacity>
          <Card.Title style={ { color: 'white'} }>{props?.item?.name}</Card.Title>
          <TouchableOpacity { ...testID('Delete') } onPress={ props.onDelete }>
            <Icon name='delete' color='white'/>
          </TouchableOpacity>
        </View>
        <Card.Divider width={2} color='white' />
        <Card.Image
          style={ { padding: 0 } }
          source={{ uri: props?.item?.img }}
        />
        <Text style={ menuStyles.lightTheme }>{props?.item?.description}</Text>
        <Text style={ menuStyles.lightTheme }>{props?.item?.price}</Text>
      </Card>
    </View>
    );
  }
}