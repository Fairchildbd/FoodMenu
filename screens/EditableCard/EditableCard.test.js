import React from 'react';
import renderer from 'react-test-renderer';
import { render, fireEvent, cleanup } from '@testing-library/react-native';
import { renderHook, act } from '@testing-library/react-hooks'
import { EditableCard, useEditing } from './EditableCard';

// If you toggle change the useState(true) and useEditing(true) in EditableCard.js
// You will see the other two tests fail, and the three tests that are failing pass.
// This is because testing-library doesn't use a virtual DOM, even though it is set up for it.
// Would need to add a virtual DOM, and assign it in this file to have the tests work as is.

describe('<EditableCard />', () => {
  afterAll(cleanup);
  const payload = {
    img: 'https://pepperoni.com',
    name: 'Extra Pepperoni Pizza',
    description: 'It has a lot of pepperoni',
    price: '$10.00'
  };
  const props = {
    handleEditedItem: jest.fn(),
    item: {
      img: 'https://api.pizzahut.io/v1/content/en-in/in-1/images/pizza/margherita.7d3b623ceb66e87fc7247fa81c6cfdc1.1.jpg?width=800',
      name: 'Cheese Pizza',
      description: 'It has a lot of cheese',
      price: '$10.25'
    },
  };
  it('can render the screen', () => {
    const tree = renderer.create(<EditableCard { ...props }/>).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('can set isEditing to true', () => {
    const { result } = renderHook(() => useEditing());
    act(() => {
      result.current.setIsEditing(true);
    });
    expect(result.current.isEditing).toBe(true)
    expect(typeof result.current.setIsEditing).toBe('function')
  });
  describe('the form', () => {
    it('can render the screen when isEditing is true', () => {
      const tree = renderer.create(<EditableCard { ...props }/>).toJSON();
      expect(tree).toMatchSnapshot();
    });
    it('can fire a submission', () => {
      const { getByTestId, getByPlaceholderText } = render(<EditableCard { ...props } />);
      fireEvent.changeText(getByTestId('Img'), payload.img);
      fireEvent.changeText(getByPlaceholderText(`${props.item.name}`), payload.name);
      fireEvent.changeText(getByPlaceholderText(`${props.item.description}`), payload.description);
      fireEvent.changeText(getByPlaceholderText(`${props.item.price}`), payload.price);
      fireEvent.press(getByTestId('Submission'), { editedItem: payload, initialItem: props.item, edited: true });
      expect(props.handleEditedItem).toHaveBeenCalledWith({ editedItem: payload, initialItem: props.item, edited: true });
    });
    it('can fire a cancellation', () => {
      const { getByTestId } = render(<EditableCard { ...props } />);
      fireEvent.press(getByTestId('Cancel'));
      expect(props.handleEditedItem).toHaveBeenCalled();
    });
  });
  describe('the card', () => {
    const props = {
      handleEditedItem: jest.fn(),
      item: {
        img: 'https://api.pizzahut.io/v1/content/en-in/in-1/images/pizza/margherita.7d3b623ceb66e87fc7247fa81c6cfdc1.1.jpg?width=800',
        name: 'Cheese Pizza',
        description: 'It has a lot of cheese',
        price: '$10.25'
      },
      onDelete: jest.fn(),
    };
    it('can fire the edit event', () => {
      const setIsEditing = jest.fn();
      const { getByTestId } = render(<EditableCard { ...props } />);
      fireEvent.press(getByTestId('Edit'));
      expect(setIsEditing).toHaveBeenCalled();
    });
    it('can fire the delete event', () => {
      const { getByTestId } = render(<EditableCard { ...props } />);
      fireEvent.press(getByTestId('Delete'));
      expect(props.onDelete).toHaveBeenCalled();
    });
  });
});