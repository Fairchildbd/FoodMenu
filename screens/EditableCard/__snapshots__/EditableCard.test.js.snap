// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EditableCard /> can render the screen 1`] = `
<View
  style={
    Object {
      "marginVertical": 10,
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "#5bd1d7",
        "borderColor": "#e1e8ee",
        "borderRadius": 10,
        "borderWidth": 1,
        "margin": 15,
        "marginBottom": 0,
        "padding": 15,
        "shadowColor": "rgba(0,0,0, .2)",
        "shadowOffset": Object {
          "height": 0,
          "width": 0,
        },
        "shadowOpacity": 1,
        "shadowRadius": 1,
      }
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "transparent",
        }
      }
    >
      <View
        style={
          Object {
            "flexDirection": "row",
            "justifyContent": "space-between",
          }
        }
      >
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
          testID="Edit"
        >
          <View
            style={
              Object {
                "overflow": "hidden",
              }
            }
          >
            <View>
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "backgroundColor": "transparent",
                    "justifyContent": "center",
                  }
                }
              >
                <Text />
              </View>
            </View>
          </View>
        </View>
        <Text
          style={
            Object {
              "color": "white",
              "fontSize": 18,
              "fontWeight": "bold",
              "marginBottom": 15,
              "textAlign": "center",
            }
          }
          testID="cardTitle"
        >
          Cheese Pizza
        </Text>
        <View
          accessible={true}
          focusable={false}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
          testID="Delete"
        >
          <View
            style={
              Object {
                "overflow": "hidden",
              }
            }
          >
            <View>
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "backgroundColor": "transparent",
                    "justifyContent": "center",
                  }
                }
              >
                <Text />
              </View>
            </View>
          </View>
        </View>
      </View>
      <View
        style={
          Object {
            "borderBottomColor": "white",
            "borderBottomWidth": 2,
            "marginBottom": 15,
          }
        }
      />
      <View
        accessibilityIgnoresInvertColors={true}
        style={
          Object {
            "backgroundColor": "transparent",
            "overflow": "hidden",
            "position": "relative",
          }
        }
      >
        <Image
          onLoad={[Function]}
          source={
            Object {
              "uri": "https://api.pizzahut.io/v1/content/en-in/in-1/images/pizza/margherita.7d3b623ceb66e87fc7247fa81c6cfdc1.1.jpg?width=800",
            }
          }
          style={
            Object {
              "bottom": 0,
              "height": 150,
              "left": 0,
              "padding": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
              "width": null,
            }
          }
          testID="RNE__Image"
          theme={
            Object {
              "colors": Object {
                "black": "#242424",
                "disabled": "hsl(208, 8%, 90%)",
                "divider": "#bcbbc1",
                "error": "#ff190c",
                "grey0": "#393e42",
                "grey1": "#43484d",
                "grey2": "#5e6977",
                "grey3": "#86939e",
                "grey4": "#bdc6cf",
                "grey5": "#e1e8ee",
                "greyOutline": "#bbb",
                "platform": Object {
                  "android": Object {
                    "error": "#f44336",
                    "grey": "rgba(0, 0, 0, 0.54)",
                    "primary": "#2196f3",
                    "searchBg": "#dcdce1",
                    "secondary": "#9C27B0",
                    "success": "#4caf50",
                    "warning": "#ffeb3b",
                  },
                  "default": Object {
                    "error": "#ff3b30",
                    "grey": "#7d7d7d",
                    "primary": "#007aff",
                    "searchBg": "#dcdce1",
                    "secondary": "#5856d6",
                    "success": "#4cd964",
                    "warning": "#ffcc00",
                  },
                  "ios": Object {
                    "error": "#ff3b30",
                    "grey": "#7d7d7d",
                    "primary": "#007aff",
                    "searchBg": "#dcdce1",
                    "secondary": "#5856d6",
                    "success": "#4cd964",
                    "warning": "#ffcc00",
                  },
                  "web": Object {
                    "error": "#ff190c",
                    "grey": "#393e42",
                    "primary": "#2089dc",
                    "searchBg": "#303337",
                    "secondary": "#ca71eb",
                    "success": "#52c41a",
                    "warning": "#faad14",
                  },
                },
                "primary": "#2089dc",
                "searchBg": "#303337",
                "secondary": "#ca71eb",
                "success": "#52c41a",
                "warning": "#faad14",
                "white": "#ffffff",
              },
            }
          }
          transition={true}
          transitionDuration={360}
        />
        <View
          accessibilityElementsHidden={true}
          importantForAccessibility="no-hide-descendants"
          pointerEvents="none"
          style={
            Object {
              "bottom": 0,
              "left": 0,
              "opacity": 1,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#bdbdbd",
                "height": 150,
                "justifyContent": "center",
                "padding": 0,
                "width": null,
              }
            }
            testID="RNE__Image__placeholder"
          />
        </View>
        <View
          style={
            Object {
              "height": 150,
              "padding": 0,
              "width": null,
            }
          }
          testID="RNE__Image__children__container"
        />
      </View>
      <Text
        style={
          Object {
            "color": "white",
            "margin": 5,
          }
        }
      >
        It has a lot of cheese
      </Text>
      <Text
        style={
          Object {
            "color": "white",
            "margin": 5,
          }
        }
      >
        $10.25
      </Text>
    </View>
  </View>
</View>
`;

exports[`<EditableCard /> the form can render the screen when isEditing is true 1`] = `
<View
  style={
    Object {
      "marginVertical": 10,
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "#5bd1d7",
        "borderColor": "#e1e8ee",
        "borderRadius": 10,
        "borderWidth": 1,
        "margin": 15,
        "marginBottom": 0,
        "padding": 15,
        "shadowColor": "rgba(0,0,0, .2)",
        "shadowOffset": Object {
          "height": 0,
          "width": 0,
        },
        "shadowOpacity": 1,
        "shadowRadius": 1,
      }
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "transparent",
        }
      }
    >
      <View
        style={
          Object {
            "flexDirection": "row",
            "justifyContent": "space-between",
          }
        }
      >
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
          testID="Edit"
        >
          <View
            style={
              Object {
                "overflow": "hidden",
              }
            }
          >
            <View>
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "backgroundColor": "transparent",
                    "justifyContent": "center",
                  }
                }
              >
                <Text />
              </View>
            </View>
          </View>
        </View>
        <Text
          style={
            Object {
              "color": "white",
              "fontSize": 18,
              "fontWeight": "bold",
              "marginBottom": 15,
              "textAlign": "center",
            }
          }
          testID="cardTitle"
        >
          Cheese Pizza
        </Text>
        <View
          accessible={true}
          focusable={false}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
          testID="Delete"
        >
          <View
            style={
              Object {
                "overflow": "hidden",
              }
            }
          >
            <View>
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "backgroundColor": "transparent",
                    "justifyContent": "center",
                  }
                }
              >
                <Text />
              </View>
            </View>
          </View>
        </View>
      </View>
      <View
        style={
          Object {
            "borderBottomColor": "white",
            "borderBottomWidth": 2,
            "marginBottom": 15,
          }
        }
      />
      <View
        accessibilityIgnoresInvertColors={true}
        style={
          Object {
            "backgroundColor": "transparent",
            "overflow": "hidden",
            "position": "relative",
          }
        }
      >
        <Image
          onLoad={[Function]}
          source={
            Object {
              "uri": "https://api.pizzahut.io/v1/content/en-in/in-1/images/pizza/margherita.7d3b623ceb66e87fc7247fa81c6cfdc1.1.jpg?width=800",
            }
          }
          style={
            Object {
              "bottom": 0,
              "height": 150,
              "left": 0,
              "padding": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
              "width": null,
            }
          }
          testID="RNE__Image"
          theme={
            Object {
              "colors": Object {
                "black": "#242424",
                "disabled": "hsl(208, 8%, 90%)",
                "divider": "#bcbbc1",
                "error": "#ff190c",
                "grey0": "#393e42",
                "grey1": "#43484d",
                "grey2": "#5e6977",
                "grey3": "#86939e",
                "grey4": "#bdc6cf",
                "grey5": "#e1e8ee",
                "greyOutline": "#bbb",
                "platform": Object {
                  "android": Object {
                    "error": "#f44336",
                    "grey": "rgba(0, 0, 0, 0.54)",
                    "primary": "#2196f3",
                    "searchBg": "#dcdce1",
                    "secondary": "#9C27B0",
                    "success": "#4caf50",
                    "warning": "#ffeb3b",
                  },
                  "default": Object {
                    "error": "#ff3b30",
                    "grey": "#7d7d7d",
                    "primary": "#007aff",
                    "searchBg": "#dcdce1",
                    "secondary": "#5856d6",
                    "success": "#4cd964",
                    "warning": "#ffcc00",
                  },
                  "ios": Object {
                    "error": "#ff3b30",
                    "grey": "#7d7d7d",
                    "primary": "#007aff",
                    "searchBg": "#dcdce1",
                    "secondary": "#5856d6",
                    "success": "#4cd964",
                    "warning": "#ffcc00",
                  },
                  "web": Object {
                    "error": "#ff190c",
                    "grey": "#393e42",
                    "primary": "#2089dc",
                    "searchBg": "#303337",
                    "secondary": "#ca71eb",
                    "success": "#52c41a",
                    "warning": "#faad14",
                  },
                },
                "primary": "#2089dc",
                "searchBg": "#303337",
                "secondary": "#ca71eb",
                "success": "#52c41a",
                "warning": "#faad14",
                "white": "#ffffff",
              },
            }
          }
          transition={true}
          transitionDuration={360}
        />
        <View
          accessibilityElementsHidden={true}
          importantForAccessibility="no-hide-descendants"
          pointerEvents="none"
          style={
            Object {
              "bottom": 0,
              "left": 0,
              "opacity": 1,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#bdbdbd",
                "height": 150,
                "justifyContent": "center",
                "padding": 0,
                "width": null,
              }
            }
            testID="RNE__Image__placeholder"
          />
        </View>
        <View
          style={
            Object {
              "height": 150,
              "padding": 0,
              "width": null,
            }
          }
          testID="RNE__Image__children__container"
        />
      </View>
      <Text
        style={
          Object {
            "color": "white",
            "margin": 5,
          }
        }
      >
        It has a lot of cheese
      </Text>
      <Text
        style={
          Object {
            "color": "white",
            "margin": 5,
          }
        }
      >
        $10.25
      </Text>
    </View>
  </View>
</View>
`;
